queue/abstract/driver/utube.lua

```diff
@@ -73,37 +93,92 @@ function method.normalize_task(self, task)
     return task and task:transform(3, 1)
 end
 
+local function put_ready(self, utube)
+    local taken = self.space.index.utube:min{state.TAKEN, utube}
+    if taken == nil or taken[2] ~= state.TAKEN then
+        local next_task = self.space.index.utube:min{state.READY, utube}
+        if next_task == nil or next_task[2] ~= state.READY then
+            return
+        end
+        pcall(self.space_ready.insert, self.space_ready, {next_task[1], utube})
+    end
+end
+
 -- put task in space
 function method.put(self, data, opts)
-    local max
-
-    -- Taking the maximum of the index is an implicit transactions, so it is
-    -- always done with 'read-confirmed' mvcc isolation level.
-    -- It can lead to errors when trying to make parallel 'put' calls with mvcc enabled.
-    -- It is hapenning because 'max' for several puts in parallel will be the same since
-    -- read confirmed isolation level makes visible all transactions that finished the commit.
-    -- To fix it we wrap it with box.begin/commit and set right isolation level.
-    -- Current fix does not resolve that bug in situations when we already are in transaction
-    -- since it will open nested transactions.
-    -- See https://github.com/tarantool/queue/issues/207
-    -- See https://www.tarantool.io/ru/doc/latest/concepts/atomic/txn_mode_mvcc/
-
-    if box.cfg.memtx_use_mvcc_engine and (not box.is_in_txn()) then
+    local commit_requirements = box.cfg.memtx_use_mvcc_engine and
+            (not box.is_in_txn())
+    if commit_requirements then
```

Comment:
If `box.cfg.memtx_use_mvcc_engine == false`, transactions will return an error in case of a yield inside a transaction. That means that, if we have (for example) two parallel `space.index:min` methods inside a transaction block, we will get an `ER_TRANSACTION_CONFLICT` error (in case of `box.cfg.memtx_use_mvcc_engine == false`).

Reply:
`space.index:min` should not yield in case of the `memtx` engine.

Reply:
I was trying to do this change, and yes, with `vinyl` I am receiving ` ER_TRANSACTION_CONFLICT: Transaction has been aborted by conflict` error.

Reply:
> I was trying to do this change, and yes, with `vinyl` I am receiving ` ER_TRANSACTION_CONFLICT: Transaction has been aborted by conflict` error.

It may happen because the `_ready` space has been not created with `vinyl` engine too:
https://github.com/tarantool/queue/blob/80f33dcd352a136c47b9dee9fdb1a4d7bf6e29c0/queue/abstract/driver/utube.lua#L68

To be honest, I don’t understand what options are used here as `space_opts`.

Reply:
In theory `vinyl` allows `yield`s inside a transaction. Try:

```Lua
box.cfg{}

box.schema.space.create('t2', {engine='vinyl'})
--box.schema.space.create('t2', {engine='memtx'})
box.space.t2:create_index('pk')

box.begin()
box.space.t2:insert({1})
require('fiber').yield()
box.space.t2.index.pk:min()
require('fiber').yield()
box.space.t2:insert({2})
require('fiber').yield()
box.commit()
```

So the problem may be caused by mixing `memtx` + `vinyl` spaces in a single transaction which is not allowed.

Reply:
Yes, this code works fine.

Reply:
For now, I suggest to implement the feature only for `memtx` and don't allow to use it with `vinyl`.

Please, create an issue to support the `vinyl` too.

Reply:
In the case of vinyl and conflict, we will probably just have to try to repeat the transaction.